### PR TITLE
Add tests for revoke permissions

### PR DESCRIPTION
**Pull request by igor-egorov**
### Description of the Change

Add acceptance tests for revoke permission command.

Tests check not only corner cases mentioned in [IR-918](https://soramitsu.atlassian.net/browse/IR-918) but also the following:
* each grantable permission is granted (separately from other grantable perms)
* test action performed using grantable permission
* grantable permission revoked
* try to perform test action again
* check that the second try was failed due to permission absence

### Benefits

Code coverage is increased.

### Possible Drawbacks 

None

### Usage Examples or Tests 

```diff
cmake -H. -Bbuild
cmake --build build --target revoke_permission_test
+or even better cmake --build build 
cd build 
ctest -R revoke_permission_test --output-on-failure
+or even better ctest --output-on-failure
```

### CODE REVIEW

File: test/integration/acceptance/grant_fixture.hpp [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+class GrantPermissionFixture : public AcceptanceFixture {
+ public:
+  using TxBuilder = TestUnsignedTransactionBuilder;
+
+  /**
+   * Creates the transaction with the user creation commands
+   * @param perms are the permissions of the user
+   * @return built tx and a hash of its payload
+   */
```
Comment (Akvinikym) [inline test/integration/acceptance/grant_fixture.hpp]: Please, add all parameters' descriptions

Comment (Akvinikym) [reply test/integration/acceptance/grant_fixture.hpp]: Also, it's unclear from the comment's head, what this function does. Please, clarify, something like: "Creates transaction, which creates a user with specified permissions"

File: test/integration/acceptance/grant_fixture.hpp [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+        .signAndAddSignature(kAdminKeypair)
+        .finish();
+  }
+
+  /**
+   * Creates two accounts with corresponding permission sets
+   * @param perm1 set of permissions for account #1
+   * @param perm2 set of permissions for account #2
+   * @return reference to ITF object with two transactions
+   */
```
Comment (Akvinikym) [inline test/integration/acceptance/grant_fixture.hpp]: `itf` parameter's description is missed

File: test/integration/acceptance/grant_fixture.hpp [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+  }
+
+  /**
+   * Forms a transaction such that creator of transaction
+   * grants permittee a permission
+   * @param creatorAccountName — a name, e.g. user
+   * @param permitteeAccountId — an id of grantee, most likely name@test
+   * @param grantPermission — any permission from the set of grantable
+   * @return a Transaction object
+   */
```
Comment (Akvinikym) [inline test/integration/acceptance/grant_fixture.hpp]: Names and number of arguments are not consistent with the function

File: test/integration/acceptance/grant_fixture.hpp [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+        .transferAsset(
+            creator_account_id, receiver_account_id, asset_id, "", amount)
+        .quorum(1)
+        .build()
+        .signAndAddSignature(creator_key)
+        .finish();
+  }
+
+  /**
+   * Transaction, that transfers standard asset from source account to receiver
```
Comment (Akvinikym) [inline test/integration/acceptance/grant_fixture.hpp]: Maybe, there's sense to clarify, what is this "standard asset"?

File: test/integration/acceptance/grant_fixture.hpp [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+        .build()
+        .signAndAddSignature(creator_key)
+        .finish();
+  }
+
+  /**
+   * Get signatories of an account (same as transaction creator)
+   * @param account_name account name which has signatories
+   * @param account_key account key which has signatories
+   * @return
```
Comment (Akvinikym) [inline test/integration/acceptance/grant_fixture.hpp]: return what?

File: test/integration/acceptance/grant_fixture.hpp [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+        .build()
+        .signAndAddSignature(account_key)
+        .finish();
+  }
+
+  /**
+   * Creates a lambda that checks query response for signatures
+   * @param signatory a keypair that has a public key to compare
+   * @param quantity required quantity of signatories
+   * @param is_contained true if the signtory is in the set
```
Comment (Akvinikym) [inline test/integration/acceptance/grant_fixture.hpp]: signtory typo

File: test/integration/acceptance/grant_fixture.hpp [hunk 7]
```diff
@@ -0,0 +1,411 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_GRANT_FIXTURE_H
```
Comment (Akvinikym) [inline test/integration/acceptance/grant_fixture.hpp]: Should be `IROHA_GRANT_FIXTURE_HPP`. Also, please, don't forget to fix the comment in the end of the file

File: test/integration/acceptance/revoke_permission_test.cpp [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+  using GrantablePermissionsTypes = ::testing::Types<AddMySignatory,
+                                                     RemoveMySignatory,
+                                                     SetMyAccountDetail,
+                                                     SetMyQuorum,
+                                                     TransferMyAssets>;
+
+  TYPED_TEST_CASE(GrantRevokeFixture, GrantablePermissionsTypes);
+
+  TYPED_TEST(GrantRevokeFixture, GrantAndRevokePermission) {
```
Comment (Akvinikym) [inline test/integration/acceptance/revoke_permission_test.cpp]: Can you, please, outline in comment, what happens in the test?

File: shared_model/interfaces/transaction_responses/tx_response.hpp [hunk 9]
```diff
@@ -53,13 +53,22 @@ namespace shared_model {
        */
       virtual const ResponseVariantType &get() const = 0;
 
+      /// Message type
+      using ErrorMessageType = std::string;
+
+      /**
+       * @return error message if present
```
[the remark this answers is not in the document]
Comment (igor-egorov) [reply shared_model/interfaces/transaction_responses/tx_response.hpp]: An empty string will be returned otherwise.

File: test/integration/acceptance/grant_fixture.hpp [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+    itf.setInitialState(kAdminKeypair)
+        .sendTx(
+            makeAccountWithPerms(kAccount1, kAccount1Keypair, perm1, kRole1))
+        .skipProposal()
+        .skipBlock()
+        .sendTx(
+            makeAccountWithPerms(kAccount2, kAccount2Keypair, perm2, kRole2))
+        .skipProposal()
+        .skipBlock();
+    return itf;
```
[the remark this answers is not in the document]
Comment (igor-egorov) [reply test/integration/acceptance/grant_fixture.hpp]: Will rework method signature to ``IntegrationTestFramework &createTwoAccounts(...)``

File: test/integration/acceptance/grant_permission_test.cpp [hunk 11]
```diff
@@ -3,387 +3,22 @@
...
-#include "framework/integration_framework/integration_test_framework.hpp"
-#include "framework/specified_visitor.hpp"
-#include "integration/acceptance/acceptance_fixture.hpp"
+#include "integration/acceptance/grantable_permissions_fixture.hpp"
 
 using namespace integration_framework;
 
 using namespace shared_model;
 using namespace shared_model::interface;
 using namespace shared_model::interface::permissions;
```
Comment (Akvinikym) [inline test/integration/acceptance/grant_permission_test.cpp]: Please, check, if all of those namespaces are still needed

File: test/integration/acceptance/grantable_permissions_fixture.cpp [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+  const auto account_id = account_name + "@" + kDomain;
+  return shared_model::proto::QueryBuilder()
+      .creatorAccountId(account_id)
+      .createdTime(getUniqueTime())
+      .queryCounter(1)
+      .getAccountDetail(account_id)
+      .build()
+      .signAndAddSignature(account_key)
+      .finish();
+}
```
Comment (Akvinikym) [inline test/integration/acceptance/grantable_permissions_fixture.cpp]: No empty line

Comment (igor-egorov) [reply test/integration/acceptance/grantable_permissions_fixture.cpp]: Thanks!

File: test/integration/acceptance/grantable_permissions_fixture.hpp [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+      const shared_model::interface::types::RoleIdType &role);
+
+  /**
+   * Creates two accounts with corresponding permission sets
+   * @param itf - initialized instance of test framework
+   * @param perm1 set of permissions for account #1
+   * @param perm2 set of permissions for account #2
+   * @return reference to ITF object with two transactions
+   */
+  integration_framework::IntegrationTestFramework &createTwoAccounts(
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.hpp]: It is not clear from the documentation, what names, domains will accounts have

File: test/integration/acceptance/grantable_permissions_fixture.hpp [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+  /**
+   * Create a transaction such that the creator grants permittee a permission
+   * @param creator_account_name - first's account name without domain
+   * @param creator_key - the keypair to sign the transaction
+   * @param permitee_account_name - a name of permittee account (receives the
+   * grantable permission)
+   * @param grant_permission - grantable permisssion to be granted
+   * @return proto::Transaction
+   */
+  shared_model::proto::Transaction accountGrantToAccount(
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.hpp]: Maybe just name it grantPermission? It is clear from parameter names that it is from account to account

File: test/integration/acceptance/grantable_permissions_fixture.hpp [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+      const shared_model::interface::types::AccountNameType
+          &creator_account_name,
+      const shared_model::crypto::Keypair &creator_key,
+      const shared_model::interface::types::AccountNameType
+          &permitee_account_name,
+      const shared_model::interface::permissions::Grantable &grant_permission);
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.hpp]: just permission for parameter name is enough since type already tells that it is grantable.

File: test/integration/acceptance/grantable_permissions_fixture.hpp [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+   * Forms a transaction such that creator of transaction revokes a permission
+   * from permittee
+   * @param creator_account_name - first's account name without domain
+   * @param creator_key - the keypair to sign the transaction
+   * @param permittee_account_name - a name of permittee account (lost the
+   * grantable permission)
+   * @param revoke_permission - grantable permission to be revoked
+   * @return proto::Transaction
+   */
+  shared_model::proto::Transaction accountRevokeFromAccount(
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.hpp]: sab

File: test/integration/acceptance/grantable_permissions_fixture.hpp [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+
+  /**
+   * Forms a transaction that either adds or removes signatory of an account
+   * @param f Add or Remove signatory function
+   * @param permitee_account_name name of account which is granted permission
+   * @param permitee_key key of account which is granted permission
+   * @param account_name account name which has granted permission to permitee
+   * @return a transaction
+   */
+  shared_model::proto::Transaction permiteeModifySignatory(
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.hpp]: just addSignatory is enough for name in my opinion

File: test/integration/acceptance/grantable_permissions_fixture.hpp [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+      TxBuilder (TxBuilder::*f)(
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.hpp]: Consider using template here. It will be clearer. Check out: https://godbolt.org/g/tNZ1Fh for an example. You can still have definition in cpp, if you explicitly instantiate it in cpp file

File: test/integration/acceptance/grantable_permissions_fixture.hpp [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+
+  /**
+   * Forms a transaction that allows permitted user to modify quorum field
+   * @param permitee_account_name name of account which is granted permission
+   * @param permitee_key key of account which is granted permission
+   * @param account_name account name which has granted permission to permitee
+   * @param quorum quorum field
+   * @return a transaction
+   */
+  shared_model::proto::Transaction permiteeSetQuorum(
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.hpp]: I think permitee in function name is redundant since you have a parameter

File: test/integration/acceptance/grantable_permissions_fixture.hpp [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+   * Forms a transaction that allows permitted user to set details of the
+   * account
+   * @param permitee_account_name name of account which is granted permission
+   * @param permitee_key key of account which is granted permission
+   * @param account_name account name which has granted permission to permitee
+   * @param key of the data to set
+   * @param detail is the data value
+   * @return a transaction
+   */
+  shared_model::proto::Transaction permiteeSetAccountDetail(
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.hpp]: sab

File: test/integration/acceptance/grantable_permissions_fixture.cpp [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+      .signAndAddSignature(kAdminKeypair)
+      .finish();
+}
+
+integration_framework::IntegrationTestFramework &
+GrantablePermissionsFixture::createTwoAccounts(
+    integration_framework::IntegrationTestFramework &itf,
+    const shared_model::interface::RolePermissionSet &perm1,
+    const shared_model::interface::RolePermissionSet &perm2) {
+  itf.setInitialState(kAdminKeypair)
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.cpp]: Won't it cause issues if initial state is already set?

File: test/integration/acceptance/grantable_permissions_fixture.cpp [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+    const shared_model::interface::types::AccountNameType &receiver_name) {
+  const auto creator_account_id = creator_name + "@" + kDomain;
+  const auto source_account_id = source_account_name + "@" + kDomain;
+  const auto receiver_account_id = receiver_name + "@" + kDomain;
+  const auto asset_id =
+      integration_framework::IntegrationTestFramework::kAssetName + "#"
+      + kDomain;
+  return TxBuilder()
+      .creatorAccountId(creator_account_id)
+      .createdTime(getUniqueTime())
```
Comment (nickaleks) [inline test/integration/acceptance/grantable_permissions_fixture.cpp]: Three lines where you set creator and time are repeated in every function, don't we have something in acceptance fixture for that, like baseTx()? If not please add it to reduce duplication

File: test/integration/acceptance/revoke_permission_test.cpp [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+      .skipBlock()
+      .sendTx(accountRevokeFromAccount(kAccount1,
+                                       kAccount1Keypair,
+                                       kAccount2,
+                                       permissions::Grantable::kSetMyQuorum))
+      .checkProposal(
+          // transaction is stateless valid
+          [](auto &proposal) { ASSERT_EQ(proposal->transactions().size(), 1); })
+      .checkBlock(
+          // transaction is not stateful valid (kAccount2 does not exist)
```
Comment (nickaleks) [inline test/integration/acceptance/revoke_permission_test.cpp]: Don't we need to check that account does not indeed exist?

Comment (nickaleks) [reply test/integration/acceptance/revoke_permission_test.cpp]: Maybe run query or something like that

File: test/integration/acceptance/grantable_permissions_fixture.cpp (hunk 21 above)
Comment (igor-egorov) [reply test/integration/acceptance/grantable_permissions_fixture.cpp]: Will move setInitialState from createTwoAccounts method.

File: test/integration/acceptance/grantable_permissions_fixture.hpp (hunk 13 above)
Comment (igor-egorov) [reply test/integration/acceptance/grantable_permissions_fixture.hpp]: Will refer to variables used for initialization

File: test/integration/acceptance/grantable_permissions_fixture.hpp (hunk 15 above)
Comment (igor-egorov) [reply test/integration/acceptance/grantable_permissions_fixture.hpp]: Discussed offline. The name `grant_permission` indicates that the permission is about to be granted. Will be left untouched.

File: test/integration/acceptance/grantable_permissions_fixture.hpp (hunk 17 above)
Comment (igor-egorov) [reply test/integration/acceptance/grantable_permissions_fixture.hpp]: Discussed offline. The name will be left untouched.

File: test/integration/acceptance/grantable_permissions_fixture.cpp (hunk 22 above)
Comment (igor-egorov) [reply test/integration/acceptance/grantable_permissions_fixture.cpp]: Thanks!
